metabase/metabase
# Handle basic/implicit actions in the action editor

Epic #27581

Adds basic/implicit actions support to the action editor. That should let people manage action's public sharing with UI + eventually customize other action properties (for now, rename and field settings editing is disabled).

Before this change, the action editor used to be tied to query actions exclusively. This PR introduces an `ActionConext` (context), providing the most up-to-date state of an action, a few properties controlling editor behavior, and callbacks for dealing with action changes. The idea is to implement an action context per action type, e.g. this PR adds query and implicit action contexts. That should hopefully make it easier to work with an editor and encapsulate action type specific behavior. For instance, query action context now completely hides the hack with creating a `Question` instance from an action to similfy working with its query.

> **Note**
> Tests for this PR are added in #28332

### To Verify

1. Go to `/model/:id/details/actions`
2. Click "Create basic actions" if your model doesn't have them yet
3. Ensure you can click the "pencil" icon next to Create/Update/Delete action
4. Ensure you can't rename an action or change its field settings
5. Ensure you can click the "gear" icon and enable/disable public sharing (ensure you have public sharing enabled for your instance)

### Demo

https://user-images.githubusercontent.com/17258145/219416869-a859a80c-dc2a-4c09-bd7f-fac49dd1a901.mp4


- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/metabase/metabase/28327)
